### PR TITLE
Sleep utilities

### PR DESCRIPTION
**Pull request by no2chem**
## Overview

Description: This PR introduces sleep utilities, which allow sleeping by time unit, and takes care of rethrowing the interrupted exception when the interrupted exception cannot be recovered from. It also adds a checkstyle error whenever Thread.sleep is called.

Why should this be merged: This is a more complete fix of #1033, and makes it a compile-time error to call Thread.sleep directly.

Related issue(s) (if applicable): #1033

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/view/replication/NeverHoleFillPolicy.java [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import org.corfudb.protocols.wireprotocol.ILogData;
 import org.corfudb.runtime.exceptions.HoleFillRequiredException;
 import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/NeverHoleFillPolicy.java]: MINOR Remove this unused import 'org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java [hunk 2]
```diff
@@ -13,6 +13,7 @@
 import org.corfudb.runtime.exceptions.unrecoverable.SystemUnavailableError;
 import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
 import org.corfudb.runtime.exceptions.WrongEpochException;
+import org.corfudb.util.Sleep;
 import org.corfudb.util.Utils;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/view/AbstractView.java]: MINOR Remove this unused import 'org.corfudb.util.Utils'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java [hunk 3]
```diff
@@ -14,6 +14,7 @@
 import org.corfudb.protocols.wireprotocol.CorfuMsgType;
 import org.corfudb.protocols.wireprotocol.JSONPayloadMsg;
 import org.corfudb.protocols.wireprotocol.VersionInfo;
+import org.corfudb.util.Sleep;
 import org.corfudb.util.Utils;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'org.corfudb.util.Utils'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: runtime/src/main/java/org/corfudb/util/Sleep.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    /** Convert this duration to the excess nanoseconds parameter used for sleep. Only
+     *  applicable for units less than a millisecond.
+     * @param duration      The duration to convert.
+     * @param milliseconds  The amount of milliseconds.
+     * @return              The number of nanoseconds leftover by subtracting milliseconds
+     *                      from duration.
+     */
+    int excessNanos(long duration, long milliseconds) {
+        return 0;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/Sleep.java]: MINOR Remove this method and declare a constant for this value. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3400)

File: runtime/src/main/java/org/corfudb/util/Sleep.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+     *
+     * @param duration      The duration to sleep.
+     */
+    @SuppressWarnings("checkstyle:ThreadSleep")
+    public void sleepUninterruptibly(long duration) {
+        try {
+            final long ms = toMillis(duration);
+            Thread.sleep(ms, excessNanos(duration, ms));
+            return;
+        } catch (InterruptedException ie) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/Sleep.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)

File: runtime/src/main/java/org/corfudb/util/Utils.java [hunk 6]
```diff
@@ -19,6 +19,7 @@
 import java.util.Set;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/Utils.java]: MINOR Remove this unused import 'java.util.concurrent.TimeUnit'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [CorfuCompileProxy.java#L27](https://github.com/CorfuDB/CorfuDB/blob/6bccb252a260b5093d1e92eecf38d084c4bc0a35/runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java#L27): Remove this unused import 'org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [Utils.java#L38](https://github.com/CorfuDB/CorfuDB/blob/6bccb252a260b5093d1e92eecf38d084c4bc0a35/runtime/src/main/java/org/corfudb/util/Utils.java#L38): Remove this unused import 'org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfu-repo** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [CorfuCompileProxy.java#L27](https://github.com/CorfuDB/CorfuDB/blob/8ac8962ca91d0ee6062c320176819d3270109759/runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java#L27): Remove this unused import 'org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **6bccb25**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1076 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/f4J5wd0MeJHCeK6h09CCAk9FqisH2U5u)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **16619fe**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1076 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/qvOKGoIZDGUh4qm25pN9libGnSfDm321)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **bde4a32**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1076 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/SUU4O9Jh356A1dRrGDswCqYHOl1yWric)

File: runtime/src/main/java/org/corfudb/util/Sleep.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    /** Convert this duration to the excess nanoseconds parameter used for sleep. Only
+     *  applicable for units less than a millisecond.
+     * @param duration      The duration to convert.
+     * @param milliseconds  The amount of milliseconds.
+     * @return              The number of nanoseconds leftover by subtracting milliseconds
+     *                      from duration.
+     */
+    int excessNanos(long duration, long milliseconds) {
+        return 0;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/Sleep.java]: MINOR Remove this method and declare a constant for this value. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3400)

File: runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java [hunk 8]
```diff
@@ -10,6 +10,7 @@
 import org.corfudb.protocols.wireprotocol.orchestrator.QueryResponse;
 import org.corfudb.runtime.CorfuRuntime;
 import org.corfudb.runtime.clients.ManagementClient;
+import org.corfudb.util.Sleep;
 import org.corfudb.util.Utils;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java]: MINOR Remove this unused import 'org.corfudb.util.Utils'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ae7da7d**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1076 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/6bTiT70SzG0ArWKvZmQVv3dM6bTKgDR3)

**Comment by no2chem** [comment]
@medhavidhawan To address technical debt, stop issues like #1033 from accidentally occurring again

File: runtime/src/main/java/org/corfudb/runtime/exceptions/unrecoverable/UnrecoverableCorfuInterruptedError.java [hunk 9]
```diff
@@ -15,7 +15,6 @@
      */
     public UnrecoverableCorfuInterruptedError(@Nonnull InterruptedException cause) {
         super(cause);
-        Thread.currentThread().interrupt();
```
[the remark this answers is not in the document]
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/exceptions/unrecoverable/UnrecoverableCorfuInterruptedError.java]: It's not supposed to reset the flag, I'll update the comment.

[the remark this answers is not in the document]
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/exceptions/unrecoverable/UnrecoverableCorfuInterruptedError.java]: Okay updated

**Comment by no2chem** [comment]
@medhavidhawan  Okay, addressed all comments.

File: runtime/src/main/java/org/corfudb/runtime/exceptions/unrecoverable/UnrecoverableCorfuInterruptedError.java [hunk 10]
```diff
@@ -4,24 +4,21 @@
...
     }
 
     /** Construct a new {@link UnrecoverableCorfuInterruptedError} with a given message
      * and root cause.
      *
-     * <p> Also resets the interrupt pending flag.
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/exceptions/unrecoverable/UnrecoverableCorfuInterruptedError.java]: but the constructor does Thread.currentThread().interrupt()

Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/exceptions/unrecoverable/UnrecoverableCorfuInterruptedError.java]: @medhavidhawan Oops missed that. Fixed.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **da0823f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1076 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/CA8do0E791znpGSChBsuFEHzDd1J8azq)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **f445b20**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1076 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/nKPQYXsmvAc4jqOtZG4Bz0MBCo5ALgFC)

**Pull request merged**